src/main/scala/io/citrine/lolo/transformers/FeatureRotator.scala

```diff
@@ -0,0 +1,221 @@
+  * @tparam T label type
+  */
+class RotatedFeatureModel[T](
+                             baseModel: Model[PredictionResult[T]],
+                             rotatedFeatures: IndexedSeq[Int],
+                             trans: DenseMatrix[Double]
+                            ) extends Model[PredictionResult[T]] {
+
+  /**
+    * Transform the inputs and then apply the base model
+    *
+    * @param inputs to apply the model to
+    * @return a RotatedFeaturePredictionResult which includes, at least, the expected outputs
+    */
+  override def transform(inputs: Seq[Vector[Any]]): RotatedFeaturePrediction[T] = {
+    val rotatedInputs = FeatureRotator.applyRotation(inputs, rotatedFeatures,  trans)
+    new RotatedFeaturePrediction(baseModel.transform(rotatedInputs), rotatedFeatures, trans)
+  }
+}
+
+/**
+  * Prediction bundling the base learner's prediction with the list of rotated features and the transformation
+  *
+  * @param baseResult
+  * @param rotatedFeatures
+  * @param trans
+  * @tparam T
+  */
+class RotatedFeaturePrediction[T](
+                                  baseResult: PredictionResult[T],
+                                  rotatedFeatures: IndexedSeq[Int],
+                                  trans: DenseMatrix[Double]
+                                 ) extends PredictionResult[T] {
+  /**
+    * Get the expected values for this prediction by delegating to baseResult
+    *
+    * @return expected value of each prediction
+    */
+  override def getExpected(): Seq[T] = baseResult.getExpected().asInstanceOf[Seq[T]]
+
+  /**
+    * Get the uncertainty of the prediction by delegating to baseResult
+    *
+    * @return uncertainty of each prediction
+    */
+  override def getUncertainty(): Option[Seq[Any]] = baseResult.getUncertainty()
+
+  /**
+    * Get the gradient or sensitivity of each prediction
+    *
+    * @return a vector of doubles for each prediction
+    */
+  override def getGradient(): Option[Seq[Vector[Double]]] = {
+    baseResult.getGradient().map { g =>
+      FeatureRotator.applyRotation(g, rotatedFeatures, trans.t).asInstanceOf[Seq[Vector[Double]]]
+    }
+  }
+
+}
+
+/**
+  * Utilities to compute and apply rotations.
+  */
+object FeatureRotator {
+
+  /**
+    * Draw a random unitary matrix from the uniform (Haar) measure.
+    *
+    * @param dimension for which to get a rotator
+    * @return unitary matrix
+    */
+  def getRandomRotation(dimension: Int): DenseMatrix[Double] = {
+    val X = DenseMatrix.rand(dimension, dimension, Gaussian(0, 1))
+    val QR(_Q, _R) = qr(X)
+    val d = signum(diag(_R))
+    val detV = d.reduce((a,b) => a*b)
+    detV * diag(d) * _Q.toDenseMatrix
+  }
+
+  /**
+    * Get list of feature indices that make sense to rotate.
+    *
+    * @param rep representative vector of features
+    * @return list of feature indices that are doubles
+    */
+  def getDoubleFeatures(rep: Vector[Any]): IndexedSeq[Int] = {
+    rep.indices.filter(i => rep(i).isInstanceOf[Double])
```

Comment:
This is fine since `rep: Vector` has O(1) indexing, but know that you could use `.zipWithIndex.filter{case (i, r) => r.isInstanceOf[Double]}.map(_._1)` to stream it if it were just `rep: Seq`.  I'm never sure a priori what will or won't work with erasure, but you might be able to do `.zipWithIndex.collect{case (i, r: double) => i}` or at least `.zipWithIndex.collect{case (i, r) if r.isInstanceOf[Double] => i}` as well.

Reply:
Thanks, so many options! I'll leave it as is, unless you have strong preference toward another variant.